blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+stopTomcat:241, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stop:295, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stopAndReleaseEmbeddedServletContainer:306, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+onClose:155, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+doClose:1014, AbstractApplicationContext (org.springframework.context.support)
+run:929, AbstractApplicationContext$2 (org.springframework.context.support)
+```
+
+Through source code analysis, it was executed by Spring's registered `ShutdownHook`.
+
+```java
+    @Override
+    public void registerShutdownHook() {
+        if (this.shutdownHook == null) {
+            // No shutdown hook registered yet.
+            this.shutdownHook = new Thread() {
+                @Override
+                public void run() {
+                    synchronized (startupShutdownMonitor) {
+                        doClose();
+                    }
+                }
+            };
+            Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+        }
+    }
+```
+
+We found that ShutdownHook will execute in the following two cases by referring the Java API documentation[2].
+
+> The Java virtual machine *shuts down* in response to two kinds of events:
+>
+> - The program *exits* normally, when the last non-daemon thread exits or when the `exit` (equivalently, [`System.exit`](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#exit-int-)) method is invoked, or
+> - The virtual machine is *terminated* in response to a user interrupt, such as typing `^C`, or a system-wide event, such as user logoff or system shutdown.
+
+1. call System.exit()
+2. Respond to external signals, such as Ctrl+C(actually sent as SIGINT signal), or `SIGTERM` signal (`kill $PID` will send `SIGTERM` signal by default)
+
+Therefore, the normal application will execute the above ShutdownHook during the stop process (except `kill -9 $PID`). Its function is not only to close the Tomcat, but also to perform other cleanup work, it is unnecessary to go into details.
+
+### Summary
+
+1. Make sure the process can't be exited by start a independency non-daemon thread loop to check the status of variable while `DubboConsumer`startup.
+2. Change the status of variable to make program exit normaly by execute ShutdownHook while `DubboConsumer` exit.
+
+### Problems
+
+In the example of DubboProvider, we see that Provider doesn't start Tomcat to provide HTTP service, then how to achieve it without exiting? We will answer this question in the next article.
+
+#### Easter Eggs/Extra/Bonus/Surprise(?)
+
+Run the following unit test in `Intellij IDEA`, create a thread to perform sleep 1000 seconds of operation, we are surprised to find that the code does not exit after the thread is executed, why is this? (The thread being created is a non-daemon thread)
```

Comment:
By running the .... which create a thread ... , we are surprised to find that the program exits with less than 1000s. Why? (...)